libsolidity/ast/ASTJsonImporter.cpp

```diff
@@ -0,0 +1,943 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @author julius <djudju@protonmail.com>
+ * @date 2019
+ *Component that imports an AST from json format to the internal format
+ */
+
+#include <libsolidity/ast/ASTJsonImporter.h>
+#include <libsolidity/ast/AsmJsonImporter.h>
+#include <liblangutil/Scanner.h>
+#include <libyul/Dialect.h>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string.hpp>
+#include <liblangutil/Token.h>
+#include <libyul/AsmParser.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <liblangutil/SourceLocation.h>
+#include <liblangutil/Exceptions.h>
+#include <liblangutil/ErrorReporter.h>
+
+
+using namespace std;
+
+using namespace dev;
+using namespace solidity;
+
+using SourceLocation = langutil::SourceLocation;
+
+template<class T>
+ASTPointer<T> ASTJsonImporter::nullOrCast(Json::Value const& _json)
+{
+	if (_json.isNull())
+		return nullptr;
+	else
+		return dynamic_pointer_cast<T>(convertJsonToASTNode(_json));
+}
+
+
+// ============ public ===========================
+
+map<string, ASTPointer<SourceUnit>> ASTJsonImporter::jsonToSourceUnit(map<string, Json::Value> const& _sourceList )
+{
+	m_sourceList = _sourceList;
+	for (auto const& src: _sourceList)
+		m_sourceLocations.emplace_back(make_shared<string const>(src.first));
+	for (auto const& srcPair: m_sourceList)
+	{
+		astAssert(!srcPair.second.isNull(), "");
+		astAssert(member(srcPair.second,"nodeType") == "SourceUnit", "The 'nodeType' of the highest node must be 'SourceUnit'.");
+		m_currentSourceName = srcPair.first;
+		m_sourceUnits[srcPair.first] =  createSourceUnit(srcPair.second, srcPair.first);
+	}
+	return m_sourceUnits;
+}
+
+// ============ private ===========================
+
+// =========== general creation functions ==============
+template <typename T, typename... Args>
+ASTPointer<T> ASTJsonImporter::createASTNode(Json::Value const& _node, Args&&... _args)
+{
+	astAssert(!member(_node, "id").isNull(), "'id'-field must not be null.");
+
+	int64_t id = _node["id"].asInt64();
+
+	if (!m_usedIDs.insert(id).second)
+		BOOST_THROW_EXCEPTION(langutil::InvalidAstError() << errinfo_comment("Found duplicate node ID!"));
+
+	auto n = make_shared<T>(
+		id,
+		createSourceLocation(_node),
+		forward<Args>(_args)...
+	);
+	return n;
+}
+
+SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _node)
+{
+	astAssert(!member(_node, "src").isNull(), "'src' can not be null");
+	string srcString = _node["src"].asString();
+	vector<string> pos;
+	boost::algorithm::split(pos, srcString, boost::is_any_of(":"));
+	if (pos.size() != 3 || int(m_sourceLocations.size()) < stoi(pos[2]))
+		BOOST_THROW_EXCEPTION(langutil::InvalidAstError() << errinfo_comment("'src'-field ill-formatted or src-index too high"));
+	int start = stoi(pos[0]);
+	int end = start + stoi(pos[1]);
```

Comment:
These look rather fishy.

Reply:
Yeah, it'd be good to properly detect, if the whole string is a valid number and throw proper errors at least...